### ISSUE TITLE
llc failure "LLVM ERROR: Stores must be a multiple of 8 bits" with WASM ASYNCIFY DEMANGLE_SUPPORT and DISABLE_EXCEPTION_CATCHING=0

### ISSUE THREAD

**Issue by lukester1975** [kripken/emscripten-fastcomp#167]
Hi

Hopefully got the right component for reporting this. I'm getting a strange llc failure with an unholy combination of WASM ASYNCIFY DEMANGLE_SUPPORT and DISABLE_EXCEPTION_CATCHING=0.

Short log output is below and I've attached the most minimal example I can trim down along with an EMCC_DEBUG=1 output in the archive. Moving everything in to one source file seems to stop it, hence the couple of cpp files and one header.

It seems to be related to usage of boost::lexical_cast. I haven't included boost in the archive, as lexical_cast seems to require most of it... 

`em++ Main.cpp Test.cpp -o Main.html -std=c++11 -s ASYNCIFY=1 -O3 -s WASM=1 -s DEMANGLE_SUPPORT=1 -s DISABLE_EXCEPTION_CATCHING=0 -I /path/to/boost_1_63_0`

fastcomp revision: dcde7043f5b94d5ad166347ab4cb9f3e3e0ff6ac

Thanks

Luke.

[wasm-demangle-exception-catching.tar.gz](https://github.com/kripken/emscripten-fastcomp/files/735879/wasm-demangle-exception-catching.tar.gz)

```
LLVM ERROR: Stores must be a multiple of 8 bits
Traceback (most recent call last):
  File "C:\Program Files\Emscripten\emscripten\incoming\\em++", line 16, in <module>
    emcc.run()
  File "C:\Program Files\Emscripten\emscripten\incoming\emcc.py", line 1647, in run
    final = shared.Building.emscripten(final, append_ext=False, extra_args=extra_args)
  File "C:\Program Files\Emscripten\emscripten\incoming\tools\shared.py", line 1745, in emscripten
    call_emscripten(cmdline)
  File "C:\Program Files\Emscripten\emscripten\incoming\emscripten.py", line 1836, in _main
    temp_files.run_and_clean(lambda: main(
  File "C:\Program Files\Emscripten\emscripten\incoming\tools\tempfiles.py", line 78, in run_and_clean
    return func()
  File "C:\Program Files\Emscripten\emscripten\incoming\emscripten.py", line 1841, in <lambda>
    DEBUG=DEBUG,
  File "C:\Program Files\Emscripten\emscripten\incoming\emscripten.py", line 1742, in main
    temp_files=temp_files, DEBUG=DEBUG)
  File "C:\Program Files\Emscripten\emscripten\incoming\emscripten.py", line 91, in emscript
    funcs, metadata, mem_init = get_and_parse_backend(infile, settings, temp_files, DEBUG)
  File "C:\Program Files\Emscripten\emscripten\incoming\emscripten.py", line 167, in get_and_parse_backend
    start_funcs = backend_output.index(start_funcs_marker)
ValueError: substring not found
```

**Comment by kripken** [comment]
Can you include boost, or give me instructions for where to get it? I want to be sure to use the exact same inputs you use to see the bug.

**Comment by lukester1975** [comment]
Sure, easiest to download it from http://www.boost.org/users/history/version_1_63_0.html

Unzip it and it'll created a boost_1_63_0 directory. Just need to add -I /path/to/boost_1_63_0 to em++, no need to build any of boost or anything like that!

Thanks

Luke.

**Comment by kripken** [comment]
Ok, I unpacked boost, but get

````
Test.cpp:5:10: fatal error: 'boost/lexical_cast.hpp' file not found
#include <boost/lexical_cast.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
````

And indeed `find -name lexical_cast.*` finds nothing - it doesn't seem to be part of boost? Maybe we are using different versions, or you have more than one installed locally?

**Comment by lukester1975** [comment]
No it's definitely there, mabye you got a bad copy somehow? Version doesn't really matter as lexical_cast has been in boost for a looong time, but I'm definitely on 1.63.0:

```
wget https://sourceforge.net/projects/boost/files/boost/1.63.0/boost_1_63_0.tar.gz
tar xzf boost_1_63_0.tar.gz
find -name lexical_cast.*
./boost_1_63_0/boost/convert/lexical_cast.hpp
./boost_1_63_0/boost/lexical_cast.hpp
./boost_1_63_0/libs/convert/doc/html/boost/cnv/lexical_cast.html
./boost_1_63_0/libs/convert/example/lexical_cast.cpp
./boost_1_63_0/libs/lexical_cast/doc/lexical_cast.qbk
```

**Comment by kripken** [comment]
Thanks, now I see that header. Very strange. The only difference is I got the `.zip` file before, not the `.tar.gz`...

Ok, now I see the problem. Looks like there is an `i2` value, and asyncify introduces it into more places, some of which the legalization code can't handle. In theory this commit could fix that, by doing legalization first,

19435d6d5e193d42098d5cde7a3b4d365efb574d

However, the problem is that asyncify is not really maintained at this point - I don't fully understand the code and I can't say I'm confident that commit is the right thing, even if it does pass the tests (which it seems to).

Possibly we should deprecate asyncify, unless someone takes an interest in the code. Have you tried the emterpreter-async option, which should be similar but better?

**Comment by lukester1975** [comment]
OK great, I'll try that patch on Monday!

Basically I'm using ASYNCIFY solely for the coroutine functions. I mentioned these with respect to emterpreter in another ticket and you said it should be possible, so I had a quick look and soon realised it wouldn't be possible by me (I know nothing of Javascript)...! If I get some time I might have another look.

Another option for me (once webassembly thread support lands) is simply to use threads as coroutines. Have done this before on other platforms without obvious coroutine/fiber support and as long as you don't have any thread-local variables it's generally OK, if not ideal.

Thanks again

**Comment by kripken** [comment]
Oh right, the coroutines... not trivial to get working, but hopefully not too hard. Hopefully we'll get to it eventually.

**Comment by lukester1975** [comment]
Yep that mod works for my real code, thanks again. I'll just keep applying it locally for now if it can't be merged.

OK I look forward to giving emterpreter coroutines a go one day. Could there be any consideration for allowing emscripten_coroutine_next() from a coroutine? It currently says:

`Run coroutine until it returns, or emscripten_yield is called. A non-zero value is returned if emscripten_yield is called, otherwise 0 is returned, and future calls of emscripten_coroutine_next on this coroutine is undefined behaviour.`

I could see relaxing that being helpful, so maybe it looks more like ucontext or Windows fibers?

```int emscripten_coroutine_next(emscripten_coroutine *previous_coroutine, emscripten_coroutine next_coroutine)```

in that it populates previousCoroutine before switching to the new one? No idea how viable that is in the emscripten world though!

Thanks

Luke.

**Comment by kripken** [comment]
I'm not familiar with the asyncify code enough to know. It could be done in the emterpreter, though. Overall, both features need someone to spend more time on them I guess.

Ditto for this patch - I'd rather not merge it in without someone having a clear idea it's the right thing to do, which means reading up on the code first.
